File: DynammicProgramming/ClimbingStairs.java
class Solution {
    int[] memo = new int[45 + 1];

    public int climbStairs(int n) {
        if (memo[n] != 0) return memo[n];

        if (n <= 1) {
            memo[n] = 1;
            return 1;
        }

        if (n == 2) {
            memo[n] = 2;
            return 2;
        }

        // num(n) = num(n - 1) + num(n - 2)
        int res = climbStairs(n - 1) + climbStairs(n - 2);
        memo[n] = res;
        return res;
    }
}

File: Python/Dynammic Programming/ClimbingStairs.py
class Solution:
    
    global memo

    def climbStairs(self, n: int) -> int:
        self.memo = (n + 1) * [-1]
        self.memo[n - 1] = 1
        self.memo[n - 2] = 2
        return self.recur(0)
        
    def recur(self, n: int) -> int:
        if n >= len(self.memo):
            return 0
        
        if self.memo[n] != -1:
            return self.memo[n]
        
        res = self.recur(n + 1) + self.recur(n + 2)
        self.memo[n] = res
        return res

# Iterative

class Solution:
    
    global memo

    def climbStairs(self, n: int) -> int:
        self.memo = (n + 1) * [-1]
        self.memo[n - 1] = 1
        self.memo[n - 2] = 2
        for i in range(n-3, -1, -1):
            self.memo[i] = self.memo[i+1] + self.memo[i+2]

        return self.memo[0]